django/db/models/sql/query.py

```diff
@@ -918,6 +918,13 @@ def join(self, join, reuse=None):
         """
         reuse = [a for a, j in self.alias_map.items()
                  if (reuse is None or a in reuse) and j == join]
+
+        if len(reuse) > 1:
+            # In some cases more than one suitable alias can be found. Join.__eq__
+            # only compares on table_name, join_field and parent_alias. Here we find
+            # the alias with a matching join_type.
+            reuse = [r for r in reuse if self.alias_map[r].join_type == join.join_type]
```

Comment:
I did this in the __eq__, but a *lot* of tests failed with differing
values. I could shift it up into the loop above and see what happens.